src/scanner.rs

```diff
@@ -0,0 +1,400 @@
+            Some(')') => Token(self.new_token(TokenType::RightParen)),
+            Some('{') => Token(self.new_token(TokenType::LeftBrace)),
+            Some('}') => Token(self.new_token(TokenType::RightBrace)),
+            Some(',') => Token(self.new_token(TokenType::Comma)),
+            Some('.') => Token(self.new_token(TokenType::Dot)),
+            Some('-') => Token(self.new_token(TokenType::Minus)),
+            Some('+') => Token(self.new_token(TokenType::Plus)),
+            Some(';') => Token(self.new_token(TokenType::Semicolon)),
+            Some('*') => Token(self.new_token(TokenType::Star)),
+            Some('!') => Token(self.if_match_next('=', TokenType::BangEqual, TokenType::Bang)),
+            Some('=') => Token(self.if_match_next('=', TokenType::EqualEqual, TokenType::Equal)),
+            Some('<') => Token(self.if_match_next('=', TokenType::LessEqual, TokenType::Less)),
+            Some('>') => {
+                Token(self.if_match_next('=', TokenType::GreaterEqual, TokenType::Greater))
+            }
+            Some('/') if self.match_next('/') => {
+                while self.peek() != Some('\n') && !self.is_at_end() {
+                    self.advance();
+                }
+                Skip
+            }
+            Some('/') if self.match_next('*') => match self.block_comment() {
+                Ok(_) => Skip,
+                Err(error) => Error(error),
+            },
+            Some('/') => Token(self.new_token(TokenType::Slash)),
+            Some(' ') => Skip,
+            Some('\t') => Skip,
+            Some('\r') => Skip,
+            Some('\n') => {
+                self.line += 1;
+                Skip
+            }
+            Some('"') => match self.string() {
+                Ok(token) => Token(token),
+                Err(error) => Error(error),
+            },
+            Some(c) => match c {
+                c if c.is_ascii_digit() => Token(self.number()),
+                c if c.is_alphabetic() || c == '_' => Token(self.identifier()),
+                c => Error(ScannerError {
+                    line: self.line,
+                    message: format!("Unexpected character {}", c),
+                }),
+            },
+        }
+    }
+
+    fn lexeme(&self) -> &'source str {
+        &self.source[self.start..self.current]
+    }
+
+    fn new_token(&self, token_type: TokenType) -> Token<'source> {
+        Token::new(token_type, self.lexeme(), self.line)
+    }
+
+    fn new_literal_token(&self, token_type: TokenType, literal: Literal) -> Token<'source> {
+        Token::new_literal(token_type, self.lexeme(), literal, self.line)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn tokenize_singles() {
+        let mut under_test = Scanner::new("(}-");
+        let tokens = under_test.scan_tokens();
+        assert!(tokens.is_ok());
+        let tokens = tokens.unwrap();
+        assert!(tokens.contains(&Token::new(TokenType::LeftParen, "(", 1)));
```

Comment:
See if there's a test framework like gtest/gmock that'll get me assert_elements_are!() or something to that effect